### DISCUSSION TITLE
Launch zenity from the pyinstaller packed app fails on some systems

### DISCUSSION THREAD

**Discussion by aFunkyBass** [pyinstaller/pyinstaller#8916] [Help] [answered]
Hello there,

this problem is driving me nuts and I'm struggling with it since the last 2 weeks.

I wrote a simple music player based on Gstreamer using customtkinter. Since the file open dialog that comes along with TKInter is really ugly, I decided to use zenity, if present on the system, and fall back to TKInter otherwise. This is used to pick up the audio file to be played.

Everything works fine on my developing box. Running the original script and the pyinstaller packed version I have no problem. If I run the executable artifact on different systems it won't launch zenity and the "stderr" grabbed by the `subprocess.run()` function reports different errors on different machines. Not to mention, on the same systems, manually launching zenity from the shell works as expected

Here is my code:

```python
import subprocess
import os

ZEN_CMD = "zenity"

def __check_zenity__() -> bool:
  try:
    subprocess.run([ZEN_CMD, "-h"], capture_output=True, text=True)
    return(True)
  except:
    return(False)


def openFileDialog(title = None, filter = None, initialdir = None):
  if(__check_zenity__()):
    return(__z_open__(**locals()))
  else:
    return(__tk_open__(**locals()))

def __z_open__(title = None, filter = None, initialdir = None):
  cmd = [ZEN_CMD, "--file-selection"]

  if(title is not None):
    cmd.append('--title')
    cmd.append(title)

  if(initialdir is None):
    initialdir = os.getcwd()

  initialdir = os.path.join(initialdir, "")

  cmd.append('--filename')
  cmd.append(initialdir)
  
  if(filter is not None):
    # example: mp3 -> MP3 files: *.mp3
    filetypes = [f"{str(x).upper()} files: *.{x}" for x in filter]
    
    # Insert all the extension together as the first available filter
    filetypes.insert(0, "Supported files: " + ' '.join(["*." + x for x in filter]))

    # Append the "All files: *" at the end of filter
    filetypes.append("All files: *")
    
    for f in filetypes:
      cmd.append("--file-filter")
      cmd.append(f)
  
  try:
    result = subprocess.run(cmd,capture_output=True, text=True)
    print(result)
    return(result.stdout.strip())
  except:
    return(None)


def __tk_open__(title = None, filter = None, initialdir = None):
  from tkinter import Tk
  import tkinter.filedialog

  filetypes = []

  filetypes.append(("Supported files:",  ' '.join(["*." + x for x in filter])))
  for x in filter:
    filetypes.append((f"{str(x).upper()} files", f"*.{x}"))

  filetypes.append(("All files", "*.*"))

  root = Tk()
  root.withdraw()
  
  f = tkinter.filedialog.askopenfilename(parent=root, initialdir = initialdir,
                                         title = title,filetypes = filetypes)

  if(len(f) == 0):
    retFile = ""
  else:
    retFile = f

  root.destroy()
  return(retFile)
``` 

This is the returning object from the subprocess call on a my developing PC:
```
CompletedProcess(args=['zenity', '--file-selection', '--title', 'Open a file', '--filename', '/home/guido/', '--file-filter', 'Supported files: *.mp3 *.wav *.flac *.aif *.ogg *.aac *.alac *.wma *.m4a', '--file-filter', 'MP3 files: *.mp3', '--file-filter', 'WAV files: *.wav', '--file-filter', 'FLAC files: *.flac', '--file-filter', 'AIF files: *.aif', '--file-filter', 'OGG files: *.ogg', '--file-filter', 'AAC files: *.aac', '--file-filter', 'ALAC files: *.alac', '--file-filter', 'WMA files: *.wma', '--file-filter', 'M4A files: *.m4a', '--file-filter', 'All files: *'], returncode=1, stdout='', stderr='')
```

This is what I get when running the program on a freshly installed Linux Mint 22 Cinnamon. I have verified that manually running the same command from the shell, open zenity with no errors
```
CompletedProcess(args=['zenity', '--file-selection', '--title', 'Open a file', '--filename', '/home/mint/', '--file-filter', 'Supported files: *.mp3 *.wav *.flac *.aif *.ogg *.aac *.alac *.wma *.m4a', '--file-filter', 'MP3 files: *.mp3', '--file-filter', 'WAV files: *.wav', '--file-filter', 'FLAC files: *.flac', '--file-filter', 'AIF files: *.aif', '--file-filter', 'OGG files: *.ogg', '--file-filter', 'AAC files: *.aac', '--file-filter', 'ALAC files: *.alac', '--file-filter', 'WMA files: *.wma', '--file-filter', 'M4A files: *.m4a', '--file-filter', 'All files: *'], returncode=127, stdout='', stderr='zenity: symbol lookup error: /lib/x86_64-linux-gnu/libatspi.so.0: undefined symbol: g_once_init_leave_pointer\n')
```

In particular this is the content of stderr:
`zenity: symbol lookup error: /lib/x86_64-linux-gnu/libatspi.so.0: undefined symbol: g_once_init_leave_pointer\n`

The rest of the app works flawlessly on every box: I can play the audio files giving them from the command line, but I can't open them when I run the app without arguments.

Thanks in advance for your help!

My dev environment:
OS: Ubuntu 20.04
Python: 3.8.10
Pyinstaller: 6.11

**Comment by rokm** [comment, marked as answer]
If you want to run a system-installed program from the frozen application, you need to ensure that the subprocess uses only system libraries. Otherwise you might get various symbol conflicts, like the one you've shown above. You do that by resetting `LD_LIBRARY_PATH` environment variable in the environment passed to your subprocess (or restoring it from `LD_LIBRARY_PATH_ORIG`, if available).

See 
https://pyinstaller.org/en/stable/common-issues-and-pitfalls.html#launching-external-programs-from-the-frozen-application
and
https://pyinstaller.org/en/stable/runtime-information.html#library-path-considerations

**Reply by aFunkyBass** [reply]
Hey @rokm 

thank you for taking the bother to reply. It is my first time using pyinstaller amd I'm a newbie.

Now that you explained the reason, it make sense to me I should have thought about it before asking, but... You know... Anyway, it works now, so thanks again.

Regards,
Guido.

**Reply by rokm** [reply]
Don't worry about it - there's a reason why the linked documentation section is called "*Common* issues and pitfalls".
